xiilib/_gunicorn/charm_state.py

```diff
@@ -0,0 +1,217 @@
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        *,
+        framework: str,
+        webserver_config: WebserverConfig,
+        is_secret_storage_ready: bool,
+        app_config: dict[str, int | str | bool] | None = None,
+        database_requirers: dict[str, DatabaseRequires] | None = None,
+        wsgi_config: dict[str, int | str] | None = None,
+        secret_key: str | None = None,
+    ):
+        """Initialize a new instance of the CharmState class.
+
+        Args:
+            framework: the framework name.
+            webserver_config: the Gunicorn webserver configuration.
+            is_secret_storage_ready: whether the secret storage system is ready.
+            app_config: User-defined configuration values for the WSGI application configuration.
+            wsgi_config: The value of the WSGI application specific charm configuration.
+            secret_key: The secret storage manager associated with the charm.
+            database_requirers: All declared database requirers.
+        """
+        self.framework = framework
+        self.service_name = self.framework
+        self.container_name = f"{self.framework}-app"
+        self.base_dir = pathlib.Path(f"/{framework}")
+        self.app_dir = self.base_dir / "app"
+        self.state_dir = self.base_dir / "state"
+        self.application_log_file = pathlib.Path(f"/var/log/{self.framework}/access.log")
+        self.application_error_log_file = pathlib.Path(f"/var/log/{self.framework}/error.log")
+        self.webserver_config = webserver_config
+        self._wsgi_config = wsgi_config if wsgi_config is not None else {}
+        self._app_config = app_config if app_config is not None else {}
+        self._is_secret_storage_ready = is_secret_storage_ready
+        self._secret_key = secret_key
+        self._database_requirers = database_requirers if database_requirers else {}
+
+    @classmethod
+    def from_charm(  # pylint: disable=too-many-arguments
+        cls,
+        charm: ops.CharmBase,
+        framework: str,
+        wsgi_config: BaseModel,
+        secret_storage: GunicornSecretStorage,
+        database_requirers: dict[str, DatabaseRequires],
+    ) -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+            framework: The WSGI framework name.
+            wsgi_config: The WSGI framework specific configurations.
+            secret_storage: The secret storage manager associated with the charm.
+            database_requirers: All database requirers object declared by the charm.
+
+        Return:
+            The CharmState instance created by the provided charm.
+        """
+        app_config = {
+            k.replace("-", "_"): v
+            for k, v in charm.config.items()
+            if not any(k.startswith(prefix) for prefix in ("flask-", "webserver-"))
```

Comment:
That's ok for now, but that's not generic enough when we'll add more frameworks. We probably need to use the framework name as reference

Reply:
Some reserved configuration prefix for example, `webserver-` are not framework names.